src/java/picard/sam/PositionBasedDownsampleSam.java

```diff
@@ -0,0 +1,361 @@
+    public double PROBABILITY = 1;
+
+    @Option(doc = "Stop after processing N reads, mainly for debugging.")
+    public long STOP_AFTER = 0;
+
+    @Option(doc = "Allow Downsampling again despite this being a bad idea with possibly unexpected results.", optional = true)
+    public boolean ALLOW_MULTIPLE_DOWNSAMPLING_DESPITE_WARNINGS = false;
+
+    @Option(doc = "Determines whether the duplicate tag should be reset since the downsampling requires re-marking duplicates.")
+    public boolean REMOVE_DUPLICATE_INFORMATION = true;
+
+    private final Log log = Log.getInstance(PositionBasedDownsampleSam.class);
+
+    private PhysicalLocation opticalDuplicateFinder;
+    private long total = 0;
+    private long kept = 0;
+    public static String PG_PROGRAM_NAME = "PositionBasedDownsampleSam";
+
+    /* max-position in tile as a function of tile. We might need to
+       look per-readgroup, but at this point I'm making the assumptions that I need to downsample a
+       sample where all the readgroups came from the same type of flowcell. */
+
+    CollectionUtil.DefaultingMap.Factory<Coord, Short> defaultingMapFactory = new CollectionUtil.DefaultingMap.Factory<Coord, Short>() {
+        @Override
+        public Coord make(final Short aShort) {
+            return new Coord();
+        }
+    };
+
+    final private Map<Short, Coord> tileCoord = new CollectionUtil.DefaultingMap<Short, Coord>(defaultingMapFactory, true);
+
+
+    final Map<Short, Histogram<Short>> xPositions = new HashMap<Short, Histogram<Short>>();
+    final Map<Short, Histogram<Short>> yPositions = new HashMap<Short, Histogram<Short>>();
+
+    public static void main(final String[] args) {
+        new PositionBasedDownsampleSam().instanceMainWithExit(args);
+    }
+
+    @Override
+    protected String[] customCommandLineValidation() {
+        final List<String> errors = new ArrayList<String>();
+
+        if (PROBABILITY < 0 || PROBABILITY > 1) {
+            errors.add("PROBABILITY must be a value between 0 and 1, found: " + PROBABILITY);
+        }
+
+        if (errors.isEmpty()) {
+            return null;
+        } else {
+            return errors.toArray(new String[errors.size()]);
+        }
+    }
+
+    @Override
+    protected int doWork() {
+
+        IOUtil.assertFileIsReadable(INPUT);
+        IOUtil.assertFileIsWritable(OUTPUT);
+
+        log.info("Checking to see if input file has been downsampled with this program before.");
+        checkProgramRecords();
+
+        opticalDuplicateFinder = new PhysicalLocation();
+
+        log.info("Starting first pass. Examining read distribution in tiles.");
+        fillTileMinMaxCoord();
+        log.info("First pass done.");
+
+        log.info("Starting second pass. Outputting reads.");
+        outputSamRecords();
+        log.info("Second pass done.");
+
+        final double finalP = kept / (double) total;
+        if (Math.abs(finalP - PROBABILITY) / (Math.min(finalP, PROBABILITY) + 1e-10) > .2) {
```

Comment:
What's the significance of the .2 magic number?  Could this be a symbolic constant?